### ISSUE TITLE
mpJwt-1.1 and mpMetrics-1.1 do not play well together

### ISSUE THREAD

**Issue by progaddict** [OpenLiberty/guide-microprofile-jwt#86]
### description
There is a service (the backend service in this case) that already uses mpJwt for authentication and authorization. MicroProfile metrics need to be added to this service. The metrics endpoint should be secured by the already existing mpJwt token authorization mechanism. The problem that it is currently not possible to do in open liberty `18.0.0.4` (or maybe possible but it is not trivial and not documented).

### steps to reproduce
*  `git clone git@github.com:progaddict/guide-microprofile-jwt.git`
*  `cd guide-microprofile-jwt`
* `git checkout add-mpMetrics-1.1`
* `mvn liberty:stop-server && mvn clean && mvn install -DskipTests liberty:start-server`
* go to `http://localhost:9090/application.jsf`
* login as `bob` (password is `bobpwd`)
* copy the bob's JWT from the table shown
* go to your favourite REST client (Postman, Insomnia etc.) and request `/metrics` from the backend service: `GET https://localhost:5051/metrics`. Use the copied bob's JWT token for authorization and ignore the SSL certificate errors (if any)
* observe the `Error 403: AuthorizationFailed` response
* observe the error in the backend server's log:
```
.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl A CWWKS9104A: Authorization failed for user bob while invoking com.ibm.ws.management.security.resource on /. The user is not granted access to any of the required roles: [Administrator, Viewer].
```

### expected outcome
since `bob` is admin he should be able to see the backend service's metrics

### attempted work around # 1
https://github.com/progaddict/guide-microprofile-jwt/commit/282a912b7fc89b25b7f28ea341aef9257b0099b7
Map bob's group `myAdmins` to `Administrator` (in the frontend's `server.xml`).
However the "Administrator" string does not appear in the JWT.
Why cannot one group be mapped to multiple security roles?

### attempted work around # 2
https://github.com/progaddict/guide-microprofile-jwt/commit/6460ff4e4a21762935b2fab9bd28a5e06f9d09ac
Define user `bob` on the backend, map him to `myAdmins` and map `myAdmins` to `admin` security role (which should come from the JWT). Additionally declare bob and group `myAdmins` to be administrators of the server:
```xml
<administrator-role>
    <user>bob</user>
    <group>myAdmins</group>
</administrator-role>
```
This does not work either.

### unsatisfactory work around that works
```xml
<mpMetrics authentication="false"/>
```
Metrics must be secured with JWT so this work around is not acceptable.

### problem
The problem boils down to the following question. Having a MP-JWT containing "admin" in the "group" claim how do I map this group to the built-in "Administrator" security role? Or, better, how do I map the "admin" MP-JWT's group to some internal group "myAdmins" and grant this internal group the server administrator role (i.e. via `<administrator-role>`)

### improvement suggestion
rather than using the built-in "Administrator" role by default (I would presume, it's hardcoded somewhere in the mpMetrics implementation), it would be nice if security roles could be setup which could access the metrics endpoint like so:
```xml
<mpMetrics>
  <allow-roles>
    <security-role name="admin"/>
  </allow-roles>
</mpMetrics>
```
or in a similar manner.

**Comment by progaddict** [comment]
## UPDATE
### attempted work around # 3
https://github.com/progaddict/guide-microprofile-jwt/commit/738e9b4d5442f10b604b9144e30607fce2cd5161
I've tried to add the "Administrator" MP-JWT's group first via additional group "mySystemAdmins" and corresponding role-group binding:
```xml
    <basicRegistry id="basic" realm="WebRealm">
        <user name="bob" password="{xor}PTA9Lyg7" />  <!-- bobpwd -->
        <user name="alice" password="{xor}PjM2PDovKDs=" />  <!-- alicepwd -->
        <user name="carl" password="{xor}PD4tMy8oOw==" />   <!-- carlpwd -->

        <group name="myAdmins">
            <member name="bob" />
        </group>

        <group name="mySystemAdmins">
            <member name="bob" />
        </group>

        <group name="myUsers">
            <member name="bob" />
            <member name="alice" />
            <member name="carl" />
        </group>
    </basicRegistry>

    <application location="${application.name}" type="war" id="${application.name}"
                 name="${application.name}" context-root="/">
        <application-bnd>
            <security-role name="admin">
                <group name="myAdmins" />
            </security-role>
            <security-role name="user">
                <group name="myUsers" />
            </security-role>
            <security-role name="Administrator">
                <group name="mySystemAdmins" />
            </security-role>
        </application-bnd>
    </application>
```
it did not work -- the JWT did not contain "Administrator" in the groups claim.

Then I've hardcoded the "Administrator" into the groups claim:
```java
  private String buildJwt(String userName, Set<String> roles) throws Exception {
      HashSet<String> jwtRoles = new HashSet<>(roles);
      jwtRoles.add("Administrator");
        return JwtBuilder.create("jwtFrontEndBuilder")
                         .claim(Claims.SUBJECT, userName)
                         .claim("upn", userName) // MP-JWT defined subject claim
                         .claim("groups", jwtRoles.toArray(new String[jwtRoles.size()])) // MP-JWT builds an array from this
                         .claim("customClaim", "customValue")
                         .buildJwt()
                         .compact();
    }
```
It didn't work either. Although the MP-JWT contained this time the "Administrator" group, this group was not mapped onto the "Administrator" security role:
```
.ibm.ws.webcontainer.security.WebAppSecurityCollaboratorImpl A CWWKS9104A: Authorization failed for user bob while invoking com.ibm.ws.management.security.resource on /. The user is not granted access to any of the required roles: [Administrator, Viewer].
```
the same log message as before

**Comment by yeekangc** [comment]
Yes, there can be improvements in this space.  @donbourne @fmhwong @chunlongliang-ibm, can you please chime in?

Using authentication filters is a path forward.  @chunlongliang-ibm, can you share an example?

FYI, @gkwan-ibm.

**Comment by gkwan-ibm** [comment]
In my understanding, the mpMetric-1.0 feature only support basic authentication mechanism but not form or custom. It should be an enhancement for mpMetric feature.

**Comment by chunlongliang-ibm** [comment]
The mpMetrics API is packaged to use management security, not to use normal web application security. So the authorization mapping is done outside web application. Here is a sample config:

```
<administrator-role>
<user-access-id>user:**put issuer_name_here**/bob</user-access-id>
</administrator-role>
```

```
<administrator-role>
<group-access-id>group:**put issuer_name_here**/myAdmins</group-access-id>
</administrator-role>
```

**Comment by chunlongliang-ibm** [comment]
The "JWT issuer name" is the token issuer name, which is used to identify the issuer, and is a URL starting with https://.

**Comment by chunlongliang-ibm** [comment]
The token issuer includes the issuer name inside JWT as "iss" claim.

**Comment by chunlongliang-ibm** [comment]
Let me know if you can not get it working. If you send me a sample JWT, I will create a working authorization mapping for you.

**Comment by progaddict** [comment]
hi @chunlongliang-ibm 
thank you for your help.

yes, sure, I'll try `<group-access-id>` setting and let you know if it works

**Comment by progaddict** [comment]
hi @chunlongliang-ibm 

I can confirm that this approach indeed works:
https://github.com/progaddict/guide-microprofile-jwt/commit/5c3746955bd6366b6a6320340bc2cb9de3d2120a

thanks a lot for your help!
the issue can be closed now.
